### ISSUE TITLE
False positive inconsistent constructor if overloads are involved.

### ISSUE THREAD

**Issue by maflAT** [microsoft/pyright#8372]
As soon as overloads are defined for `__new__` or `__init__` reportInconsistentConstructor will be triggered, even if the overloads are identical.

Code sample in [pyright playground](https://pyright-play.net/?strict=true&locale=en&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDKApgDbAA0UYAbkSCWAIYAmAsAFAcDGJjAznygBhAFwcoEqAAEadBi3GTmRYFAD6alEQDuGgBQ8%2BASigBaAHyFSwEVAB0DiQGIoCOCCRoAFjBGeUuEQA2iBEOHgAkihcYCh8SHwwRCgwAPK09EzMALqKEsqqGlq6agYkxmaWxGRi7JL1UKEwAK4gKFCGekYcedKymQp1SirqaqhIMPp81iYWUABysUS2DnbOru6ePn5oAaEhYbgwUTFxCUkp6XJZuUP5Ixrjk6XTZLOWi1q2UL31CPx8XouACyCQgjBgXC8UAARkQYNoiMkoPFdhDWkQqIVNDoNFBGChmKMnnjUB1eAIoAAiIRUoGEfzo0JY4koCakwRUvSvGzCIx0u5QFwERktZlgbHFDnUsp8WzwBDBarAKRCbL8qAABTgvGiRD0oXCx2isXiiWSMCEppgIGaXBguCMQA)

```python
from typing import Self, overload

class C:
    @overload
    def __new__(cls) -> Self: ...  # pyright:ignore[reportInconsistentOverload]
    def __new__(cls) -> Self:
        return cls()

    @overload
    def __init__(self) -> None: ...  # pyright:ignore[reportInconsistentOverload]
    def __init__(self) -> None:  
        pass
    # Mismatch between signature of __new__ and __init__ in class "C"
    # Signature of __init__ is "(self: C)"
    # Signature of __new__ is "(cls: type[Self@C])" Pylance(reportInconsistentConstructor)
```

**Comment by erictraut** [comment]
Thanks for the bug report. Yes, this is behaving incorrectly. 

When pyright sees an overloaded `__init__` or `__new__`, it is supposed to fall back on the overload's implementation to verify constructor consistency. The code was indeed doing this, but it was not binding the implementation to the class, which means that the `self` and `cls` parameters for `__init__` and `__new__` (respectively) will not match.

This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.372.
